target/linux/ath79/dts/qca9531_glinet_gl-ar300m-lite.dts

```diff
@@ -7,7 +9,9 @@
 	model = "GL.iNet GL-AR300M-Lite";
 };
 
-// GL-AR300M-Lite has different LED colors than the non-Lite version
+/delete-node/ &nand_flash;
```

Comment:
Generally delete-node should be avoid in dts because It's a bit confusing. Just add nand node in devices with nand instead.

Reply:
In this special case (the [GL-AR300M series](https://www.gl-inet.com/products/gl-ar300m/)) the GL-AR300M was the initial model with 16M NOR and 128M NAND. Later they introduced the GL-AR300M-Lite (w/ one ethernet, now discontinued) and the GL-AR300M16 (w/ two ethernet) editions which are NOR only models.

With the `delete-node` it's in line more with the product series, IMHO.

With this PR the base model is able to boot both flash.

Reply:
@xabolcs Don't take me wrong, but I do not think we should build the structure of the DTS files based on the chronology of the models.

With these devices, it will be relatively simple to have two DTSIs, one with nor only and another one just adding the nand_flash, as 981213 suggested. No matter what the history of the devices is, this IMO will the least-confusing way to represent the hardware of those devices.

Reply:
Yup, a restructure of the DTSIs are needed also if this PR would support the newer GL-AR300M16 model.

Reply:
@xabolcs
Can you give some details about that? In this PR, the AR300M16 was just replacing the AR300M-NOR in the generic target.
I wonder whether we couldn't just add the AR300M16 right away without touching the SPI-NAND topic. So, if I understand this correctly, one would now just add the M16 as an _additional_ device, and then this PR would later move the old ar300m-nor to the nand-target?

Reply:
@adrianschmutzler 
Ohh! Sorry for the noise! I was too focused on the Lite path here.

Reply:
I thought the M16 was already supported  in master and/or the M16 and NAND versions are indistinguishable currently since the only difference is "NAND that is not there"?

Reply:
@Entropy512 If I understood it correctly, at the moment you would just take the ar300m-nor image. I just created a support patch adding it based on the code here, and it will just be a clone of the ar300m-nor with different name then:
https://github.com/adrianschmutzler/openwrt/commits/m16

Reply:
The AR300M16 support could be separated from the SPI-NAND topic.

The GL-AR300M model (with 16M NOR + 128M NAND flash) is able to boot from both flash chip. So it's very nice feature that there is a "(NOR)" and "(NAND)" notation (e.g. in the `dmesg`) and one could differentiate what flash was booted from.

Sure the AR-300M booting to NOR is basically an AR-300M16. ;)

Reply:
@adrianschmutzler does the Overview in the PR's description help you to understand the big picture of this PR? OK that could be out of sync with the commits.

Reply:
I think I've become aware of the big picture in the meantime. I just try to reorganize this PR in a way that it is easier to understand for outsiders and to separate different steps more clearly from each other. That's why I also try to pull out some changes which might be dealt with independent of the big picture.

For the two subjects at hand:
1. In my opinion, using delete-node or adding stuff in different DTSIs is a matter of taste here. However, based on my tasted the latter is preferred over the former.
2. After having assembled the separate commit for the AR300M16, I realized why this is pointless as a separate commit... ;-)

Reply:
The AR300M series are all the same PCB and boot loader. The differences are how the board is populated by GL.iNet to meet either a specific customer's need or market demand.
* Leave off the NAND chip
* Use a single Ethernet socket instead of a dual
* Change the colors of the LEDs
* Add in RF connectors for external antennas
* ...

In keeping with ease of maintenance, and aligned with the best practices of reuse and only declaring something once in the code, the DTS/DTSI structure reflects this.

Especially as the NAND-aware firmware versions can read both the NAND and the NOR flash, the partition maps need to be consistent between them, now and in the future.

Duplicating them needlessly, or splitting them out into another DTSI seems contrary to maintainability and robustness. 

delete-node is part of the DTS spec and is clearly documented both in the spec and in descriptions of how to work with DTS files. Nowhere have I found its use admonished

https://devicetree-specification.readthedocs.io/en/latest/source-language.html
https://developer.toradex.com/device-tree-customization#delete-properties-or-nodes
https://elinux.org/Device_Tree_Source_Undocumented

On the question of the GL-AR300M16 and GL-AR300M-Lite, they are on the "generic" ath79 target as:

1. Primarily, they don't need the significant space load of SPI-NAND, UBI in their kernels or that of the NAND-related firmware utilities
2. It helps prevents "accidentally" upgrading, for example a GL-AR300M16 to a NAND-based firmware

(Yes, one _could_ run a GL-AR300M16 on the NAND-kernel, NOR-flash firmware, firmware-size penalty. Given the number of complaints about 4/32 devices, I'd prefer not to overburden the device, especially for those that want to fill its 16 MB flash.)